src/main/java/com/epam/ta/reportportal/core/item/impl/DeleteTestItemHandlerImpl.java

```diff
@@ -80,36 +78,65 @@ public DeleteTestItemHandlerImpl(TestItemRepository testItemRepository, LogRepos
+	public OperationCompletionRS deleteTestItem(Long itemId, ReportPortalUser.ProjectDetails projectDetails, ReportPortalUser user) {
 		TestItem item = testItemRepository.findById(itemId)
 				.orElseThrow(() -> new ReportPortalException(ErrorType.TEST_ITEM_NOT_FOUND, itemId));
-
 		Launch launch = launchRepository.findById(item.getLaunchId())
 				.orElseThrow(() -> new ReportPortalException(ErrorType.LAUNCH_NOT_FOUND, item.getLaunchId()));
 
-		validate(item, launch, reportPortalUser, projectDetails);
+		validate(item, launch, user, projectDetails);
 		Optional<TestItem> parent = ofNullable(item.getParent());
 
 		testItemRepository.deleteById(item.getItemId());
 
-		logIndexer.cleanIndex(projectDetails.getProjectId(), logRepository.findIdsByTestItemId(item.getItemId()));
-
 		launch.setHasRetries(launchRepository.hasRetries(launch.getId()));
-
 		parent.ifPresent(p -> p.setHasChildren(testItemRepository.hasChildren(p.getItemId(), p.getPath())));
 
+		logIndexer.cleanIndex(projectDetails.getProjectId(), logRepository.findIdsByTestItemId(item.getItemId()));
 		eventPublisher.publishEvent(new DeleteTestItemAttachmentsEvent(item.getItemId()));
 
-		return new OperationCompletionRS("Test Item with ID = '" + itemId + "' has been successfully deleted.");
+		return COMPOSE_DELETE_RESPONSE.apply(item.getItemId());
 	}
 
 	@Override
-	public List<OperationCompletionRS> deleteTestItem(Long[] ids, ReportPortalUser.ProjectDetails projectDetails,
-			ReportPortalUser reportPortalUser) {
-		return Stream.of(ids).map(it -> deleteTestItem(it, projectDetails, reportPortalUser)).collect(toList());
+	public List<OperationCompletionRS> deleteTestItems(Collection<Long> ids, ReportPortalUser.ProjectDetails projectDetails,
+			ReportPortalUser user) {
+		List<TestItem> items = testItemRepository.findAllById(ids);
+		List<Launch> launches = launchRepository.findAllById(items.stream().map(TestItem::getLaunchId).collect(Collectors.toSet()));
+
+		Map<Long, List<TestItem>> launchItemMap = items.stream().collect(Collectors.groupingBy(TestItem::getLaunchId));
+		launches.forEach(launch -> launchItemMap.get(launch.getId()).forEach(item -> validate(item, launch, user, projectDetails)));
+
+		List<Long> willBeDeletedExplicitly = new ArrayList<>(items.size());
+		List<Long> shouldBeDeletedCascaded = new ArrayList<>(items.size());
+
+		Map<Integer, List<TestItem>> grouppedByLevel = items.stream()
+				.collect(Collectors.groupingBy(it -> it.getType().getLevel(), TreeMap::new, Collectors.toList()));
+
+		grouppedByLevel.forEach((key, value) -> value.stream().filter(item -> !shouldBeDeletedCascaded.contains(item.getItemId())).forEach(item -> {
+			willBeDeletedExplicitly.add(item.getItemId());
+			shouldBeDeletedCascaded.addAll(testItemRepository.selectAllDescendantsIds(item.getPath()));
+		}));
+
+		testItemRepository.deleteAllByItemIdIn(willBeDeletedExplicitly);
+
+		launches.forEach(it -> it.setHasRetries(launchRepository.hasRetries(it.getId())));
+		items.stream()
```

Comment:
probably, you can go through all "rootItems"/"willBeDeleteExplicitly", take all of its' non null parents and set them "hasDescedants" = false;